Add `icon` parameter to st.expander

snehankekre opened the pull request:
## Describe your changes

Adds an `icon` kwarg to `st.expander` to optionally support single-character emojis or material icons.

- Reused the existing `icon` proto field
- Replaced the custom check and error icons in `st.status` with `DynamicIcon` material icons


## GitHub Issue Link (if applicable)

N/A

## Testing Plan

- [x] Python unit tests
- [x] FE JS test
- [x] E2E test 

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

snehankekre committed:
[Prototype] Add `icon` parameter to st.expander

snehankekre committed:
Reuse ExpanderIcon for custom icon rendering

snehankekre committed:
Make custom icons respect FE theme

snehankekre committed:
Use existing icon proto for expander icon param

snehankekre committed:
Use DynamicIcon for status error and check icons in Expander.tsx

snehankekre committed:
Don't specify margin, padding type in computeSpacingStyle for StyledDynamicIcon

snehankekre committed:
Merge remote-tracking branch 'origin' into snehan/prototype/expander-icon

snehankekre committed:
Fix DynamicIconProps import in test

snehankekre committed:
Make size optional again in DynamicIconProps

snehankekre committed:
Remove erroneous stExpanderIconError data-testid from user provided expander icons

snehankekre committed:
Remove default margin, padding assignment in DynamicIconDispatcher

snehankekre committed:
Merge remote-tracking branch 'origin' into snehan/prototype/expander-icon

snehankekre committed:
Typo: `button` -> `expander` in docstring

snehankekre committed:
Python unit tests for valid/invalid emoji/material icons

snehankekre committed:
FE tests for emoji and material icons in expanders

snehankekre committed:
Merge remote-tracking branch 'origin' into snehan/prototype/expander-icon

snehankekre committed:
Remove unused imports from DynamicIcon.tsx: IconSize, ThemeColor

snehankekre committed:
E2E test for expander icon

snehankekre committed:
Upload st_expander-with-icon snapshots

lukasmasuch commented on lib/streamlit/elements/layouts.py:
```diff
@@ -385,7 +386,12 @@ def tab_proto(label: str) -> BlockProto:
+        self,
+        label: str,
+        expanded: bool = False,
+        icon: str | None = None,
```
nit: I think it should be added as keyword-only parameter (that's our current policy with all new parameters):

```suggestion
        *,
        icon: str | None = None,
```

snehankekre replied:
Oops, I forgot this detail which was in the spec. Fixed 👍

lukasmasuch commented on e2e_playwright/st_expander.py:
```diff
@@ -67,3 +67,7 @@ def update_value():
 )
 collapsed_long.write("I am already collapsed")
+
+expander_icon = st.expander("Expand me!", icon=":material/bolt:").write(
```
nit: maybe also add one test with an emoji icon.

snehankekre replied:
Added another test with an emoji icon

lukasmasuch commented on e2e_playwright/st_expander_test.py:
```diff
@@ -35,13 +35,14 @@ def test_expander_displays_correctly(
     assert_snapshot(expander_elements.nth(3), name="st_expander-with_input")
     assert_snapshot(expander_elements.nth(4), name="st_expander-long_expanded")
     assert_snapshot(expander_elements.nth(5), name="st_expander-long_collapsed")
+    assert_snapshot(expander_elements.nth(6), name="st_expander-with-icon")
```
super nit:

```suggestion
    assert_snapshot(expander_elements.nth(6), name="st_expander-with_icon")
```

snehankekre replied:
Argh! Still happened after you reminded me in my last PR. Fixed now 👍

lukasmasuch commented on frontend/lib/src/components/shared/Icon/DynamicIcon.tsx:
```diff
@@ -38,32 +38,35 @@ function parseIconPackEntry(iconName: string): IconPackEntry {
-      return <MaterialFontIcon pack={pack} iconName={icon} {...props} />
+      return (
+        <StyledDynamicIcon iconValue={icon} {...props}>
+          <MaterialFontIcon pack={pack} iconName={icon} {...props} />
+        </StyledDynamicIcon>
+      )
     case "emoji":
     default:
-      return <EmojiIcon {...props}>{icon}</EmojiIcon>
+      return (
+        <StyledDynamicIcon iconValue={icon} {...props}>
```
I think it might be good to add `data-testid` to both of these options, e.g.: `stIconEmoji` and `stIconMaterial` and use these test ids in the tests. Or maybe we actually set the `testid` from `DynamicIconProps` here, at the moment it seems to be unused, or?

snehankekre replied:
I've added data-testids for `stIconEmoji` and `stIconMaterial` and used these in the tests 👍 
But I also did the following and not sure the last part is needed:
- For st.status's check and error icons, we used to add `data-testid`: `stExpanderIconCheck` and `stExpanderIconError`. I continue passing them to `<DynamicIcon>`
- When the icon value is neither check nor error, I added a generic data-testid `stExpanderIcon`. Not sure if this is really needed or if I should remove it and not specify one.

lukasmasuch replied:
Hmm, I think both solutions would be fine... but I'm slightly favoring removing this distinction and only have one  test id to have lower tech debt. I think the related unit tests aren't that important since we are anyways testing this via e2e snapshot tests as well.

snehankekre replied:
I've removed the distinction and only used on test id `stExpanderIcon`.

lukasmasuch commented on frontend/lib/src/components/shared/Icon/DynamicIcon.tsx:
```diff
@@ -38,32 +38,35 @@ function parseIconPackEntry(iconName: string): IconPackEntry {
   return { pack: iconPack, icon: iconNameInPack }
 }
 
-export interface DynamicIconProps {
```
I think it's better to keep the `DynamicIconProps` and add another interface specific for `StyledDynamicIcon` in `styled-components,`, e.g., `StyledDynamicIconProps.` While the props are quite similar in this case, its still two different components ... combining it will make it more complicated later if we add something that's only relevant to the `DynamicIconProps` level or the other way around.

snehankekre replied:
> its still two different components ... combining it will make it more complicated later if we add something that's only relevant to the DynamicIconProps level or the other way around

I hadn't considered that case 😅 Thanks for calling it out. Fixed!

lukasmasuch commented on e2e_playwright/st_expander_test.py:
```diff
@@ -94,3 +95,18 @@ def test_expander_session_state_set(app: Page):
+    expander_with_icon = expanders[5]
+    expect(expander_with_icon.locator("svg")).to_be_visible()
+
+    icon = expander_with_icon.locator("span").last
```
nit: it might be better to use a test-id if we have defined a good one. See comment above

snehankekre replied:
I've fixed that in the tests :)

lukasmasuch commented on frontend/lib/src/components/elements/Expander/Expander.tsx:
```diff
@@ -75,26 +72,30 @@ export const ExpanderIcon = (props: ExpanderIconProps): ReactElement => {
         aria-hidden="true"
         data-testid="stExpanderIconCheck"
         {...iconProps}
       />
     )
   } else if (icon === "error") {
     return (
-      <StyledIcon
-        as={ErrorOutline}
-        color={"inherit"}
+      <DynamicIcon
+        iconValue=":material/error:"
+        color="inherit"
         aria-hidden="true"
         data-testid="stExpanderIconError"
         {...iconProps}
       />
     )
   }
 
```
One consideration here could be to fully remove these two icons here and instead just set `:material/check:` or `:material/error:` in the backend for icon instead of `check` or `error`. Having that here feels a bit redundant since its basically doing the same as just using `:material/check:`. We would just need to remove or adapt the unit test that uses the testIDs here and check for the icon value instead.

snehankekre replied:
Ahh, why didn't I think of that. Really good point! I've set the material directive in the backend now.

snehankekre committed:
Incorporate @LukasMasuch feedback

snehankekre committed:
Replace check and error with material directives in testing/v1/element_tree.py and layouts_test.py

snehankekre committed:
Upload missing snapshots

snehankekre committed:
Merge remote-tracking branch 'origin' into snehan/prototype/expander-icon

lukasmasuch commented on frontend/lib/src/components/elements/Expander/Expander.tsx:
```diff
@@ -41,15 +37,16 @@ import {
  *
  * If the icon is "spinner", it will render a spinner icon.
  * If the icon is "check", it will render a check icon.
  * If the icon is "error", it will render an error icon.
```
nit: I think these two comments can be removed with the recent change

snehankekre replied:
I've removed the two comments 👍

lukasmasuch commented on frontend/lib/src/components/elements/Expander/Expander.tsx:
```diff
@@ -73,29 +76,19 @@ export const ExpanderIcon = (props: ExpanderIconProps): ReactElement => {
+    <DynamicIcon
+      color="inherit"
+      iconValue={icon}
+      aria-hidden={statusIconTestIds[icon] ? "true" : undefined}
```
nit: I don't think we have to differentiate here, I think we can even just remove the `aria-hidden` completely since the new icons are probably more accessible since they actually contain descriptive text.

snehankekre replied:
Removed `aria-hidden` completely ✅

lukasmasuch commented on frontend/lib/src/components/shared/Icon/styled-components.ts:
```diff
@@ -82,6 +82,35 @@ export const StyledIcon = styled("span", {
+  margin?: string
+  padding?: string
+  testid?: string
+  color?: ThemeColor
```
I think we can remove a couple of properties here that are not used for styling: `iconValue` and `testid`

snehankekre replied:
Done 😄

lukasmasuch commented on frontend/lib/src/components/elements/Expander/Expander.tsx:
```diff
@@ -73,29 +76,19 @@ export const ExpanderIcon = (props: ExpanderIconProps): ReactElement => {
+      color="inherit"
+      iconValue={icon}
+      aria-hidden={statusIconTestIds[icon] ? "true" : undefined}
+      data-testid={statusIconTestIds[icon] || "stExpanderIcon"}
```
It think we better should use the `testid` prop defined in DynamicIcon here instead. Using other props here is working since we are routing it through to the styled component, but that's a very unusual pattern (maybe we should even remove this part). It is better torelyy on the props that are explicitly defined in DynamicIcon.

snehankekre replied:
Good call! I saw the prop forwarding being done on `StyledIcon` and didn't question if it was unusual. But I've removed that in `StyledDynamicIcon` and made it rely on props that are explicitly defined in DynamicIcon 👍 Also, made the switch from `data-testid` to the `testid` prop defined in DynamicIcon

lukasmasuch reviewed (approved):
LGTM 👍 A couple more comments.

snehankekre committed:
Merge branch 'develop' into snehan/prototype/expander-icon

snehankekre committed:
Use props explicitly defined in DynamicIcon. Don't route via styled component

snehankekre committed:
Use only one testid `stExpanderIcon` in expander tests

snehankekre committed:
Delete offending snapshots and regenerate in CI

snehankekre committed:
Update snapshots

snehankekre commented:
@LukasMasuch Sorry for the long delay! I've finally addressed all the remaining comments 😄 If there aren't any other blocking issues, we could merge and get this into the 1.36 release.

snehankekre committed:
Merge remote-tracking branch 'origin' into snehan/prototype/expander-icon

snehankekre commented:
IMO #8889 doesn't subsume this change. With a dedicated `icon` for expanders, we're able to style the icons better than inline in labels

lukasmasuch reviewed (approved):
Yep, let's merge it 👍

lukasmasuch commented on frontend/lib/src/components/shared/Icon/Material/styled-components.ts:
```diff
@@ -23,6 +23,7 @@ export interface StyledMaterialIconProps {
   margin: string
   padding: string
   color: string
+  testid?: string
```
Is this property needed?

snehankekre replied:
Nope! Good catch. Removed on 044bff3

lukasmasuch commented on frontend/lib/src/components/shared/Icon/styled-components.ts:
```diff
@@ -82,6 +82,32 @@ export const StyledIcon = styled("span", {
 })
 
+export interface StyledDynamicIconProps {
+  iconValue: string
```
I think iconValue and testid can also be removed here since it's not used by the styled component

snehankekre replied:
😬 I missed these again, not needed. Removed on 044bff3.

lukasmasuch commented on e2e_playwright/st_expander_test.py:
```diff
@@ -94,3 +96,21 @@ def test_expander_session_state_set(app: Page):
+def test_expander_renders_icon(app: Page):
+    """Test that an expander renders a material icon and an emoji icon."""
+    main_container = app.get_by_test_id("stAppViewBlockContainer")
+    main_expanders = main_container.get_by_test_id("stExpander")
```
nit: I recently added a better method in e2e_playwright.shared.app_utils to get an expander by label (`get_expander`). This might be a more stable way compared to getting the first expander in the app view container.

snehankekre replied:
Sweet! I've used it in this test on 1866a63

snehankekre committed:
Remove unused props from StyledDynamicIcon and StyledMaterialIcon

snehankekre committed:
Use `get_expander` util in `test_expander_renders_icon`

snehankekre committed:
Upload st_expander icon snapshots

jmckenna commented:
As mentioned in a previous issue, it would be great if these "add..." pull requests could include a screen capture of the proposed change in action, would really help the users, that follow the changelog.
